server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java

```diff
@@ -1752,4 +1761,495 @@ public void testSubobjectsFalseDocWithEmptyObject() throws IOException {
+                        "type": "ip"
+                      }
+                    }
+                  }
+                ]
+              }
+            }
+            """;
+        String docJson = """
+            {
+                "one100_ip": "11.11.11.120",
+                "iptwo": "10.10.10.10",
+                "threeip": "12.12.12.12"
+            }
+            """;
+
+        MapperService mapperService = createMapperService(mapping);
+        ParsedDocument parsedDoc = mapperService.documentMapper().parse(source(docJson));
+        merge(mapperService, dynamicMapping(parsedDoc.dynamicMappingsUpdate()));
+        LuceneDocument doc = parsedDoc.rootDoc();
+
+        assertEquals(InetAddressPoint.class, doc.getField("one100_ip").getClass());
+        Mapper fieldMapper = mapperService.documentMapper().mappers().getMapper("one100_ip");
+        assertNotNull(fieldMapper);
+        assertEquals("ip", fieldMapper.typeName());
+
+        assertEquals(InetAddressPoint.class, doc.getField("iptwo").getClass());
+        fieldMapper = mapperService.documentMapper().mappers().getMapper("iptwo");
+        assertNotNull(fieldMapper);
+        assertEquals("ip", fieldMapper.typeName());
+
+        assertNotEquals(InetAddressPoint.class, doc.getField("threeip").getClass());
+        fieldMapper = mapperService.documentMapper().mappers().getMapper("threeip");
+        assertNotNull(fieldMapper);
+        assertEquals("text", fieldMapper.typeName());
+    }
+
+    public void testMatchWithArrayOfFieldNamesMixingGlobsAndRegex() throws IOException {
+        String mapping = """
+            {
+              "_mappings": {
+                "dynamic_templates": [
+                  {
+                    "test": {
+                      "match": ["one*", ".*two$", "^xyz.*"],
+                      "mapping": {
+                        "type": "ip"
+                      }
+                    }
+                  }
+                ]
+              }
+            }
+            """;
+        String docJson = """
+            {
+                "oneip": "11.11.11.120",
+                "iptwo": "10.10.10.10",
+                "threeip": "12.12.12.12"
+            }
+            """;
+
+        MapperService mapperService = createMapperService(mapping);
+        ParsedDocument parsedDoc = mapperService.documentMapper().parse(source(docJson));
+        merge(mapperService, dynamicMapping(parsedDoc.dynamicMappingsUpdate()));
+        LuceneDocument doc = parsedDoc.rootDoc();
+
+        assertEquals(InetAddressPoint.class, doc.getField("oneip").getClass());
+        Mapper fieldMapper = mapperService.documentMapper().mappers().getMapper("oneip");
+        assertNotNull(fieldMapper);
+        assertEquals("ip", fieldMapper.typeName());
+
+        // this one will not match and be an IP field because it was specified with a regex but match_pattern is implicit "simple"
+        assertNotEquals(InetAddressPoint.class, doc.getField("iptwo").getClass());
```

Comment:
Yes, but that is existing behavior. The validate method in `DynamicTemplate.parse` tries to catch these, but it only catches the invalid pattern when `match_pattern=regex`. For example, see the test: https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplateParseTests.java#L132

And doing ad-hoc testing on the main branch:

```
  "mappings": {
    "dynamic_templates": [
      {
        "nameynames": {
          "match_pattern": "regex",
          "match": "*name",
          "unmatch": "*middle",
          "mapping": {
            "type": "keyword"
          }
        }
      }
    ]
  }
```
fails with:
```
    "type": "mapper_parsing_exception",
    "reason": "Failed to parse mapping: Pattern [*name] of type [regex] is invalid. Cannot create dynamic template [nameynames].",
```

but:
```
      {
        "nameynames": {
          "match_pattern": "simple",
          "match": ".*name$",
          "unmatch": "^.*middle.*",
          "mapping": {
            "type": "keyword"
          }
        }
      }
```
works with no problems (until you try to actually put documents in with the mapping, of course).

Reply:
> I wonder if in a follow up we ought to emit a warning if you have a pattern that 'looks like' a regex, but haven't explicitly set match_type?

Probably a good idea from a user perspective, but that would be new functionality. Should we add it here or create a new issue/ticket for it?

We could probably add a method to Regex.java. There is already this function
```
   public static boolean isSimpleMatchPattern(String str) {
        return str.indexOf('*') != -1;
    }
```

but that's not enough for this scenario.

[This code](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/regex/Regex.java#L133) indicates that the only allowed patterns for "simple" are `"xxx*", "*xxx", "*xxx*" and "xxx*yyy"`, so we could check for things like `.*` and `^xxx` and `yyy$` and chars like `[`, `]` and `|` and maybe parens and then run it through [REGEX.matches()](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java#L40) and if that does NOT fail, then it appears to be a valid actual regex, not a simple pattern and then give a warning to users.

My vote would be to add that in a separate ticket so that if it causes an issue for some reason, we don't have to back out the core functionality of this ticket.

But I'm happy to take a stab at it either way. Let me know what you think.

Reply:
Yes, let's do this in a follow-up, no need to complicate this PR.